File: src/loss.py
import numpy as np

from layer import Layer 


def convert_to_hot_one(label: int, num_labels: int) -> np.ndarray:
    hot_one: np.ndarray = np.zeros(shape = num_labels)
    hot_one[label] = 1.0
    return hot_one

EPSILON = 1e-5

class CategoricalCrossEntropy(Layer):

    def __init__(self):
        self.target_classes: list = []

    def forward(self, input: np.ndarray, labels: np.ndarray) -> np.ndarray:
        self.forward_input = np.clip(input, EPSILON, 1.0 - EPSILON)
        num_classes: int = len(np.unique(labels))
        self.target_classes = np.array([convert_to_hot_one(label, num_classes) for label in labels])
        self.forward_output = - self.target_classes * np.log(self.forward_input)
        return self.forward_output
    
    def backward(self, input: np.ndarray) -> np.ndarray:
        self.backward_output = - self.target_classes / input
        return self.backward_output / len(input)

File: src/Loss.cxx
module;

import std;
import tensor;
import config;
import dense_layer;

export module loss;

namespace Loss
{
    static constexpr auto OneHotEncoding = []<int R, int C> [[nodiscard]] (const auto& labels)
    {
        auto result = Tensor<float, { R, C }>{ 0.0f };

        for (auto i : std::ranges::views::iota(0, R)) {
            result.FillAt(i, labels.At(i, 0), 1.0f);
        }

        return result;
    };

    export [[nodiscard]] constexpr auto Accuracy(const auto* input, const auto& labels)
    {
        const auto rows         = labels.Shape().first;
        const auto predictions  = input->ArgMax();
        auto correctPredictions = 0;

        for (auto i : std::ranges::views::iota(0, rows)) {
            if (labels.At(i) == predictions.At(i)) {
                ++correctPredictions;
            }
        }

        return static_cast<float>(correctPredictions) / static_cast<float>(rows);
    }

    export class CategoricalCrossEntropy final
    {
    public:
        template<LayerParams prevLayer>
        [[nodiscard]] constexpr auto Forward (
            const Tensor<float, { prevLayer.Inputs, prevLayer.Neurons }>& input,
            const Tensor<int, { prevLayer.Inputs, 1 }>& trueLabels
        )
        {
            m_trueLabels     = trueLabels;
            auto confidences = Tensor<float, { prevLayer.Inputs, 1 }>{};

            // TODO: Change to std::views::enumerate when available
            for (auto i = 0; const auto& row : input.Data()) {
                confidences.FillAt(i, 0, row.at(trueLabels.At(i)));
                ++i;
            }

            static constexpr auto minFloat = std::numeric_limits<float>::min();

            confidences.Clip(minFloat, 1.0f - minFloat); // Clip to prevent log(0.0f)
            confidences.Log();
            confidences.Negative();

            return confidences.Mean();
        }

        template<LayerParams nextLayer>
        [[nodiscard]] constexpr auto Backward (
            const Tensor<float, { nextLayer.Inputs, nextLayer.Neurons }>& gradients
        )
        {
            auto labels = OneHotEncoding.operator()<Config::BatchSize, Config::NumClasses>(m_trueLabels);
            auto output = labels / gradients;
            output.Negative();
            output.MultiplyEachElementBy( 1.0f / static_cast<float>(nextLayer.Inputs) );
            return output;
        }

    private:
        Tensor<int, { Config::BatchSize, 1 }> m_trueLabels;
    };
}
